File: sdk/typescript/src/actions/groups.ts
import { jsonRequest, listParams, pathId, requestInit, withParams } from "../internal/utils.js";
import type {
  CreateGroupMembershipRequest,
  CreateGroupRequest,
  Group,
  GroupListParams,
  GroupMembership,
  GroupMembershipResponse,
  GroupResponse,
  ListParams,
  ListResponse,
  UpdateGroupMembershipRequest,
  UpdateGroupRequest,
  UUID,
} from "../types.js";
import { searchableListParams } from "./params.js";
import type { KivalClientBase, WithSignal } from "./types.js";

/** Parameters for {@link createGroup}. */
export type CreateGroupParameters = WithSignal<{
  input: CreateGroupRequest;
}>;

/** Return type for {@link createGroup}. */
export type CreateGroupReturnType = Group;

/** Creates a group. */
export function createGroup(
  client: KivalClientBase,
  parameters: CreateGroupParameters,
): Promise<CreateGroupReturnType> {
  const { input, signal } = parameters;
  return client
    .requestJson<GroupResponse>("/groups", jsonRequest("POST", input, signal))
    .then((response) => response.group);
}

/** Parameters for {@link listGroups}. */
export type ListGroupsParameters = WithSignal<GroupListParams>;

/** Return type for {@link listGroups}. */
export type ListGroupsReturnType = ListResponse<Group>;

/** Lists groups. */
export function listGroups(
  client: KivalClientBase,
  parameters: ListGroupsParameters = {},
): Promise<ListGroupsReturnType> {
  const params = searchableListParams(parameters);
  params.set("status", parameters.status ?? "active");
  return client.requestJson<ListGroupsReturnType>(
    withParams("/groups", params),
    requestInit({}, parameters.signal),
  );
}

/** Parameters for {@link getGroup}. */
export type GetGroupParameters = WithSignal<{
  groupId: UUID;
}>;

/** Return type for {@link getGroup}. */
export type GetGroupReturnType = Group;

/** Gets a group by ID. */
export function getGroup(
  client: KivalClientBase,
  parameters: GetGroupParameters,
): Promise<GetGroupReturnType> {
  const { groupId, signal } = parameters;
  return client
    .requestJson<GroupResponse>(`/groups/${pathId(groupId)}`, requestInit({}, signal))
    .then((response) => response.group);
}

/** Parameters for {@link updateGroup}. */
export type UpdateGroupParameters = WithSignal<{
  groupId: UUID;
  input: UpdateGroupRequest;
}>;

/** Return type for {@link updateGroup}. */
export type UpdateGroupReturnType = Group;

/**
 * Updates a group.
 *
 * Omitting `description` leaves it unchanged, `null` clears it, and a string replaces it.
 */
export function updateGroup(
  client: KivalClientBase,
  parameters: UpdateGroupParameters,
): Promise<UpdateGroupReturnType> {
  const { groupId, input, signal } = parameters;
  return client
    .requestJson<GroupResponse>(`/groups/${pathId(groupId)}`, jsonRequest("PATCH", input, signal))
    .then((response) => response.group);
}

/** Parameters for {@link archiveGroup}. */
export type ArchiveGroupParameters = WithSignal<{
  groupId: UUID;
}>;

/** Return type for {@link archiveGroup}. */
export type ArchiveGroupReturnType = Group;

/** Archives a group. */
export function archiveGroup(
  client: KivalClientBase,
  parameters: ArchiveGroupParameters,
): Promise<ArchiveGroupReturnType> {
  const { groupId, signal } = parameters;
  return client
    .requestJson<GroupResponse>(
      `/groups/${pathId(groupId)}/archive`,
      requestInit({ method: "POST" }, signal),
    )
    .then((response) => response.group);
}

/** Parameters for {@link unarchiveGroup}. */
export type UnarchiveGroupParameters = WithSignal<{
  groupId: UUID;
}>;

/** Return type for {@link unarchiveGroup}. */
export type UnarchiveGroupReturnType = Group;

/** Unarchives a group. */
export function unarchiveGroup(
  client: KivalClientBase,
  parameters: UnarchiveGroupParameters,
): Promise<UnarchiveGroupReturnType> {
  const { groupId, signal } = parameters;
  return client
    .requestJson<GroupResponse>(
      `/groups/${pathId(groupId)}/unarchive`,
      requestInit({ method: "POST" }, signal),
    )
    .then((response) => response.group);
}

/** Parameters for {@link listGroupMemberships}. */
export type ListGroupMembershipsParameters = WithSignal<
  ListParams & {
    groupId: UUID;
  }
>;

/** Return type for {@link listGroupMemberships}. */
export type ListGroupMembershipsReturnType = ListResponse<GroupMembership>;

/** Lists active memberships in a group. */
export function listGroupMemberships(
  client: KivalClientBase,
  parameters: ListGroupMembershipsParameters,
): Promise<ListGroupMembershipsReturnType> {
  const { groupId, ...options } = parameters;
  return client.requestJson<ListGroupMembershipsReturnType>(
    withParams(`/groups/${pathId(groupId)}/memberships`, listParams(options)),
    requestInit({}, options.signal),
  );
}

/** Parameters for {@link createGroupMembership}. */
export type CreateGroupMembershipParameters = WithSignal<{
  groupId: UUID;
  input: CreateGroupMembershipRequest;
}>;

/** Return type for {@link createGroupMembership}. */
export type CreateGroupMembershipReturnType = GroupMembership;

/** Creates a group membership. */
export function createGroupMembership(
  client: KivalClientBase,
  parameters: CreateGroupMembershipParameters,
): Promise<CreateGroupMembershipReturnType> {
  const { groupId, input, signal } = parameters;
  return client
    .requestJson<GroupMembershipResponse>(
      `/groups/${pathId(groupId)}/memberships`,
      jsonRequest("POST", input, signal),
    )
    .then((response) => response.membership);
}

/** Parameters for {@link updateGroupMembership}. */
export type UpdateGroupMembershipParameters = WithSignal<{
  groupId: UUID;
  membershipId: UUID;
  input: UpdateGroupMembershipRequest;
}>;

/** Return type for {@link updateGroupMembership}. */
export type UpdateGroupMembershipReturnType = GroupMembership;

/** Updates an active group membership's role. */
export function updateGroupMembership(
  client: KivalClientBase,
  parameters: UpdateGroupMembershipParameters,
): Promise<UpdateGroupMembershipReturnType> {
  const { groupId, membershipId, input, signal } = parameters;
  return client
    .requestJson<GroupMembershipResponse>(
      `/groups/${pathId(groupId)}/memberships/${pathId(membershipId)}`,
      jsonRequest("PATCH", input, signal),
    )
    .then((response) => response.membership);
}

/** Parameters for {@link revokeGroupMembership}. */
export type RevokeGroupMembershipParameters = WithSignal<{
  groupId: UUID;
  membershipId: UUID;
}>;

/** Return type for {@link revokeGroupMembership}. */
export type RevokeGroupMembershipReturnType = GroupMembership;

/** Revokes a group membership. */
export function revokeGroupMembership(
  client: KivalClientBase,
  parameters: RevokeGroupMembershipParameters,
): Promise<RevokeGroupMembershipReturnType> {
  const { groupId, membershipId, signal } = parameters;
  return client
    .requestJson<GroupMembershipResponse>(
      `/groups/${pathId(groupId)}/memberships/${pathId(membershipId)}/revoke`,
      requestInit({ method: "POST" }, signal),
    )
    .then((response) => response.membership);
}

File: sdk/rust/src/client/groups.rs
//! Group API operations.

use reqwest::Method;
use uuid::Uuid;

use crate::{
    ClientError, CreateGroupMembershipRequest, CreateGroupRequest, Group, GroupListParams,
    GroupMembership, GroupMembershipResponse, GroupResponse, KivalClient, ListParams, ListResponse,
    Transport, UpdateGroupMembershipRequest, UpdateGroupRequest,
    client::transport::append_list_params,
};

impl<S> KivalClient<S>
where
    S: Transport,
{
    /// Lists groups.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn list_groups(
        &self,
        params: &GroupListParams,
    ) -> Result<ListResponse<Group>, ClientError> {
        self.require_api_key()?;

        let mut url = self.api_url("/groups")?;
        append_list_params(
            &mut url,
            &ListParams { limit: params.limit, cursor: params.cursor.clone() },
        );
        {
            let mut pairs = url.query_pairs_mut();
            pairs.append_pair("status", params.status.as_str());
            if let Some(q) = &params.q {
                pairs.append_pair("q", q);
            }
        }

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Gets a group by ID.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn get_group(&self, group_id: Uuid) -> Result<Group, ClientError> {
        self.require_api_key()?;

        let request = self.request(&Method::GET, &format!("/groups/{group_id}"))?;
        let response = self.send_json::<GroupResponse>(request).await?;

        Ok(response.group)
    }

    /// Creates a group.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn create_group(&self, request: CreateGroupRequest) -> Result<Group, ClientError> {
        self.require_api_key()?;

        let request = self.request(&Method::POST, "/groups")?.json(&request);
        let response = self.send_json::<GroupResponse>(request).await?;

        Ok(response.group)
    }

    /// Updates a group.
    ///
    /// `description` is tri-state: `None` leaves it unchanged, `Some(None)` clears it, and
    /// `Some(Some(value))` sets it.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn update_group(
        &self,
        group_id: Uuid,
        request: UpdateGroupRequest,
    ) -> Result<Group, ClientError> {
        self.require_api_key()?;

        let request = self.request(&Method::PATCH, &format!("/groups/{group_id}"))?.json(&request);
        let response = self.send_json::<GroupResponse>(request).await?;

        Ok(response.group)
    }

    /// Archives a group.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn archive_group(&self, group_id: Uuid) -> Result<Group, ClientError> {
        self.require_api_key()?;

        let request = self.request(&Method::POST, &format!("/groups/{group_id}/archive"))?;
        let response = self.send_json::<GroupResponse>(request).await?;

        Ok(response.group)
    }

    /// Unarchives a group.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn unarchive_group(&self, group_id: Uuid) -> Result<Group, ClientError> {
        self.require_api_key()?;

        let request = self.request(&Method::POST, &format!("/groups/{group_id}/unarchive"))?;
        let response = self.send_json::<GroupResponse>(request).await?;

        Ok(response.group)
    }

    /// Lists active memberships in a group.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn list_group_memberships(
        &self,
        group_id: Uuid,
        params: &ListParams,
    ) -> Result<ListResponse<GroupMembership>, ClientError> {
        self.require_api_key()?;

        let mut url = self.api_url(&format!("/groups/{group_id}/memberships"))?;
        append_list_params(&mut url, params);

        let request = self.request_url(&Method::GET, url)?;
        self.send_json(request).await
    }

    /// Creates a group membership.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn create_group_membership(
        &self,
        group_id: Uuid,
        request: CreateGroupMembershipRequest,
    ) -> Result<GroupMembership, ClientError> {
        self.require_api_key()?;

        let request =
            self.request(&Method::POST, &format!("/groups/{group_id}/memberships"))?.json(&request);
        let response = self.send_json::<GroupMembershipResponse>(request).await?;

        Ok(response.membership)
    }

    /// Updates an active group membership's role.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn update_group_membership(
        &self,
        group_id: Uuid,
        membership_id: Uuid,
        request: UpdateGroupMembershipRequest,
    ) -> Result<GroupMembership, ClientError> {
        self.require_api_key()?;

        let request = self
            .request(&Method::PATCH, &format!("/groups/{group_id}/memberships/{membership_id}"))?
            .json(&request);
        let response = self.send_json::<GroupMembershipResponse>(request).await?;

        Ok(response.membership)
    }

    /// Revokes a group membership.
    ///
    /// # Errors
    ///
    /// Returns [`ClientError::ApiKeyRequired`] if no API key is configured.
    pub async fn revoke_group_membership(
        &self,
        group_id: Uuid,
        membership_id: Uuid,
    ) -> Result<GroupMembership, ClientError> {
        self.require_api_key()?;

        let request = self.request(
            &Method::POST,
            &format!("/groups/{group_id}/memberships/{membership_id}/revoke"),
        )?;
        let response = self.send_json::<GroupMembershipResponse>(request).await?;

        Ok(response.membership)
    }
}
